presto-main/src/main/java/com/facebook/presto/operator/scalar/ScalarFunctionImplementation.java

```diff
@@ -124,4 +149,80 @@ public boolean isReturnValueAsParameter()
     {
         return returnValueSliceLength;
     }
+
+    public synchronized ScalarFunctionImplementation ensureReturnValueAsReturn()
+    {
+        if (!returnValueAsParameter) {
+            return this;
+        }
+        if (returnValuesAsReturnImplementation == null) {
+            MethodHandle newMethodHandle = generateReturnAsReturnMethodHandler(this.methodHandle, returnValueSliceLength.get(), instanceFactory.isPresent());
+            returnValuesAsReturnImplementation = new ScalarFunctionImplementation(
+                    nullable,
+                    nullableArguments,
+                    newMethodHandle,
+                    instanceFactory,
+                    deterministic,
+                    false,
+                    Optional.empty());
+        }
+        return returnValuesAsReturnImplementation;
+    }
+
+    private MethodHandle generateReturnAsReturnMethodHandler(MethodHandle originalMethodHandle, int returnValueSliceLength, boolean usesInstance)
+    {
+        MethodType originalMethodType = originalMethodHandle.type();
+        CallSiteBinder callSiteBinder = new CallSiteBinder();
+        ClassDefinition classDefinition = new ClassDefinition(
+                a(PUBLIC, FINAL),
+                makeClassName("ScalarImplementation"),
+                type(Object.class));
+
+        List<Parameter> newParameters = newArrayList();
+        for (int i = 0; i < originalMethodType.parameterCount(); ++i) {
+            if (i != (usesInstance ? 1 : 0)) {
+                newParameters.add(arg("p" + i, originalMethodType.parameterType(i)));
+            }
+        }
+        MethodDefinition methodDefinition = classDefinition.declareMethod(a(PUBLIC, STATIC), GENERATED_METHOD_NAME, type(Slice.class), newParameters);
+
+        Variable returnVariable = methodDefinition.getScope().declareVariable(Slice.class, "returnValue");
+        BytecodeBlock methodBody = methodDefinition.getBody();
+
+        methodBody.append(constantInt(returnValueSliceLength))
+                .invokeStatic(Slices.class, "allocate", Slice.class, int.class)
+                .putVariable(returnVariable);
+
+        if (usesInstance) {
+            methodBody.append(newParameters.get(0));
+        }
+        methodBody.getVariable(returnVariable);
+        for (int i = usesInstance ? 1 : 0; i < newParameters.size(); ++i) {
+            methodBody.append(newParameters.get(i));
+        }
+        methodBody.append(invoke(callSiteBinder.bind(originalMethodHandle), "originalMethod"));
+        methodBody.getVariable(returnVariable);
+        methodBody.ret(Slice.class);
+
+        Class<?> generatedClass = defineClass(classDefinition, Object.class, callSiteBinder.getBindings(), getClass().getClassLoader());
+        return lookupNewMethodHandle(generatedClass);
```

Comment:
Isn't it possible that generated classes will have name conflicts with each other?